docs/csharp/language-reference/compiler-messages/ref-modifiers-errors.md

```diff
@@ -0,0 +1,366 @@
+- **CS8168** - *Cannot return local by reference because it is not a ref local*
+- **CS8169** - *Cannot return a member of local variable by reference because it is not a ref local*
+- **CS8345** - *Field or auto-implemented property cannot be of type unless it is an instance member of a `ref struct`.*
+- **CS8351** - *Branches of a `ref` conditional operator cannot refer to variables with incompatible declaration scopes*
+- **CS8374** - *Cannot ref-assign source has a narrower escape scope than destination.*
+- **CS9077** - *Cannot return a parameter by reference through a `ref` parameter; it can only be returned in a return statement*
+- **CS9078** - *Cannot return by reference a member of parameter through a `ref` parameter; it can only be returned in a return statement*
+- **CS9079** - *Cannot ref-assign source to destination because source can only escape the current method through a return statement.*
+- **CS9096** - *Cannot ref-assign source to destination because source has a wider value escape scope than destination allowing assignment through destination of values with narrower escapes scopes than source.*
+- **CS9109** - *Cannot use `ref`, `out`, or `in` primary constructor parameter inside an instance member*
+- **CS9110** - *Cannot use primary constructor parameter that has ref-like type inside an instance member*
+
+Warnings:
+
+- **CS9085** - *This ref-assigns source to destination but source has a narrower escape scope than destination.*
+- **CS9086** - *The branches of the ref conditional operator refer to variables with incompatible declaration scopes*
+- **CS9087** - *This returns a parameter by reference but it is not a `ref` parameter*
+- **CS9089** - *This returns by reference a member of parameter that is not a `ref` or `out` parameter*
+- **CS9091** - *This returns local by reference but it is not a ref local*
+- **CS9092** - *This returns a member of local by reference but it is not a ref local*
+- **CS9093** - *This ref-assigns source to destination but source can only escape the current method through a return statement.*
+- **CS9094** - *This returns a parameter by reference through a `ref` parameter; but it can only safely be returned in a return statement*
+- **CS9095** - *This returns by reference a member of parameter through a `ref` parameter; but it can only safely be returned in a return statement*
+- **CS9097** - *This ref-assigns source to destination but source has a wider value escape scope than destination allowing assignment through destination of values with narrower escapes scopes than source.*
+
+The compiler uses static analysis to determine if the referent is valid at all points where the reference variable can be used. You need to refactor code so that the referent remains valid at all locations where the reference variable may refer to it. For details on the rules for ref safety, see the C# standard on [ref safe contexts](~/_csharpstandard/standard/variables.md#972-ref-safe-contexts).
```

Comment:
```suggestion
The compiler uses static analysis to determine if the referent is valid at all points where the reference variable can be used. You need to refactor code so that the referent remains valid at all locations where the reference variable might refer to it. For details on the rules for ref safety, see the C# standard on [ref safe contexts](~/_csharpstandard/standard/variables.md#972-ref-safe-contexts).
```